### PR TITLE
Enhancement/optimize edge all predicate

### PR DESCRIPTION
**Pull request by czpmango**
## What type of PR is this?
- [ ] bug
- [ ] feature
- [x] enhancement

## What problem(s) does this PR solve?
Push edge all predicate down to the storage layer to optimize performance.

## Checklist:
Tests:
- [ ] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

### CODE REVIEW

**Review by jievince — DISMISSED** [review]
LGTM

**Comment by yixinglu** [comment]
plz solve the conflicts of this PR

**Comment by czpmango** [comment]
> plz solve the conflicts of this PR

Ok.

File: src/graph/executor/query/TraverseExecutor.cpp [hunk 1]
```diff
@@ -62,8 +62,12 @@ Status TraverseExecutor::buildRequestVids() {
     auto vidType = SchemaUtil::propTypeToValueType(metaVidType.get_type());
     for (; iter->valid(); iter->next()) {
       const auto& vid = src->eval(ctx(iter));
-      DCHECK_EQ(vid.type(), vidType)
```
Comment (Shylock-Hg) [inline src/graph/executor/query/TraverseExecutor.cpp:65]: Why delete this?

File: src/graph/optimizer/CMakeLists.txt [hunk 2]
```diff
@@ -57,8 +58,9 @@ nebula_add_library(
     rule/PushLimitDownScanEdgesAppendVerticesRule.cpp
     rule/PushTopNDownIndexScanRule.cpp
     rule/PushLimitDownScanEdgesRule.cpp
-    rule/PushFilterDownTraverseRule.cpp
+    rule/PushFilterThroughAppendVerticesRule.cpp
     rule/RemoveAppendVerticesBelowJoinRule.cpp
+    rule/EmbedEdgeAllPredIntoTraverseRule.cpp
```
Comment (Shylock-Hg) [inline src/graph/optimizer/CMakeLists.txt:63]: What's the difference from `PushFilterDownTraverse` rule?

File: src/graph/executor/query/TraverseExecutor.cpp (hunk 1 above)
Comment (czpmango) [reply src/graph/executor/query/TraverseExecutor.cpp:65]: crash

**Review by jievince — DISMISSED** [review]

File: src/graph/executor/query/TraverseExecutor.cpp (hunk 1 above)
Comment (Shylock-Hg) [reply src/graph/executor/query/TraverseExecutor.cpp:65]: Should check why vid type is unexpected.

File: tests/tck/features/optimizer/EmbedEdgeAllPredIntoTraverseRule.feature [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+      | [99]             | 29   |
+      | [99]             | 31   |
+      | [99]             | 33   |
+    And the execution plan should be:
+      | id | name           | dependencies | profiling data | operator info                             |
+      | 7  | Project        | 15           |                |                                           |
+      | 15 | Filter         | 11           |                | {"condition": "((n.player.age>0)==true)"} |
+      | 11 | AppendVertices | 14           |                |                                           |
+      | 14 | Filter         | 13           |                | {"condition": "(size($e)>0)"}             |
+      | 13 | Traverse       | 1            |                | {"edge filter": "(*.likeness>90)"}        |
```
Comment (yixinglu) [inline tests/tck/features/optimizer/EmbedEdgeAllPredIntoTraverseRule.feature:158]: why the predicate is sometimes in `edge filter` and sometimes in `filter` fields of `Traverse`?

File: tests/tck/features/optimizer/EmbedEdgeAllPredIntoTraverseRule.feature [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+      | 44 | HashLeftJoin   | 15,43        |                |                                                                              |
+      | 15 | Dedup          | 14           |                |                                                                              |
+      | 14 | Project        | 35           |                |                                                                              |
+      | 35 | HashInnerJoin  | 53,50        |                |                                                                              |
+      | 53 | Filter         | 52           |                | {"condition": "(id($friend)!=\"Tony Parker\")"}                              |
+      | 52 | AppendVertices | 59           |                |                                                                              |
+      | 59 | Filter         | 60           |                | {"condition": "(id($person)==\"Tony Parker\")"}                              |
+      | 60 | Traverse       | 2            |                | {"filter": "((like.likeness>0) AND (like.likeness>1))"}                      |
```
Comment (yixinglu) [inline tests/tck/features/optimizer/EmbedEdgeAllPredIntoTraverseRule.feature:273]: we could merge the condition into `like.likeness>1` later.

**Review by yixinglu — COMMENTED** [review]
good job!

maybe you should test some scenarios which could not push the `all` predicates down, such as:

```
all( i in e where i.likeness>95 AND true);
all(i in e WHERE i.likeness>95 AND v.player.name =="");
all(i in e WHERE i.likeness>95 OR i.likeness < 50);
...
```

**Comment by czpmango** [comment]
> all(i in e WHERE i.likeness>95 AND v.player.name =="");

ACK.

File: src/graph/executor/query/TraverseExecutor.cpp (hunk 1 above)
Comment (czpmango) [reply src/graph/executor/query/TraverseExecutor.cpp:65]: ACK.

File: tests/tck/features/optimizer/EmbedEdgeAllPredIntoTraverseRule.feature (hunk 3 above)
Comment (czpmango) [reply tests/tck/features/optimizer/EmbedEdgeAllPredIntoTraverseRule.feature:158]: 0 step filter cannot push to storage layer.

File: tests/tck/features/optimizer/EmbedEdgeAllPredIntoTraverseRule.feature (hunk 4 above)
Comment (czpmango) [reply tests/tck/features/optimizer/EmbedEdgeAllPredIntoTraverseRule.feature:273]: Yep. Expression normalization is not optimized yet.

**Review by yixinglu — APPROVED** [review]

**Review by jievince — APPROVED** [review]

**Pull request merged**